README.md
# flight-simulator

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar flight-simulator-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-05-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-05-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/flight-simulator/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/flight-simulator/compare/0.1.0...0.1.1


test/flight_simulator/core_test.clj
(ns flight-simulator.core-test
  (:require [clojure.test :refer :all]
            [flight-simulator.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/flight_simulator/sequential.clj
(ns flight-simulator.sequential
  (:require [clojure.string]
            [clojure.pprint]
            #_[input-simple :as input]
            [flight-simulator.input-random :as input]))

(def logger (agent nil))
(defn log [& msgs] (send logger (fn [_] (apply println msgs))))
;(defn log [& msgs] nil)

(def flights
  "All flights are encapsulated in a single atom in this implementation.
  You are free to change this to a more appropriate mechanism."
  (atom []))

(defn initialize-flights [initial-flights]
  "Set `flights` atom to the `initial-flights`."
  (reset! flights initial-flights))

(defn print-flights [flights]
  "Print `flights`."
  (letfn [(pricing->str [pricing]
            (->> pricing
              (map (fn [[p a t]] (clojure.pprint/cl-format nil "$~3d: ~3d ~3d" p a t)))
              (clojure.string/join ", ")))]
    (doseq [{:keys [id from to pricing]} flights]
      (println (clojure.pprint/cl-format nil "Flight ~3d from ~a to ~a: ~a"
        id from to (pricing->str pricing))))))

(defn- update-pricing [flight factor]
  "Updated pricing of `flight` with `factor`."
  (update flight :pricing
    #(map (fn [[p a t]] [(* p factor) a t]) %)))

(defn start-sale [carrier]
  "Sale: all flights of `carrier` -20%."
  (log "Start sale for" carrier "!")
  (swap! flights
    (fn [old-flights]
      (vec (map
             (fn [flight]
               (if (= (:carrier flight) carrier)
                 (update-pricing flight 0.80)
                 flight))
             old-flights)))))

(defn end-sale [carrier]
  "End sale: all flights of `carrier` +25% (inverse of -20%)."
  (log "End sale for" carrier "!")
  (swap! flights
    (fn [old-flights]
      (vec (map
             (fn [flight]
               (if (= (:carrier flight) carrier)
                 (update-pricing flight 1.25)
                 flight))
             old-flights)))))

(defn sort-pricing [pricing]
  "Sort `pricing` from lowest to highest price."
  (sort-by first pricing))

(defn filter-pricing-with-n-seats [pricing seats]
  "Get `pricing` for which there are at least `seats` empty seats available."
  (filter #(>= (second %) seats) pricing))

(defn lowest-available-price [flight seats]
  "Returns the lowest price in `flight` for which at least `seats` empty seats
  are available, or nil if none found."
  (-> (:pricing flight)                 ; [[price available taken]]
    (filter-pricing-with-n-seats seats)
    (sort-pricing)
    (first)                             ; [price available taken]
    (first)))                           ; price

(defn- find-flight [flights customer]
  "Find a flight in `flights` that is on the route and within the budget of
  `customer`. If a flight was found, returns {:flight flight :price price},
  else returns nil."
  (let [{:keys [_id from to seats budget]}
          customer
        flights-and-prices
          ; flights that are on the route and within budget, and their price
          (for [f flights
                :when (and (= (:from f) from) (= (:to f) to))
                :let [lowest-price (lowest-available-price f seats)]
                :when (and (some? lowest-price) (<= lowest-price budget))]
            {:flight f :price lowest-price})
        cheapest-flight-and-price
          (first (sort-by :price flights-and-prices))]
    cheapest-flight-and-price))

(defn- book [flight price seats]
  "Updates `flight` to book `seats` at `price`."
  (update flight :pricing
    (fn [pricing]
      (for [[p a t] pricing]
        (if (= p price)
          [p (- a seats) (+ t seats)]
          [p a t])))))

(defn- process-customer [flights customer]
  "Try to book a flight from `flights` for `customer`, returning the updated
  flight if found, or nil if no suitable flight was found."
  (if-let [{:keys [flight price]} (find-flight flights customer)]
    (let [updated-flight (book flight price (:seats customer))]
      (log "Customer" (:id customer) "booked" (:seats customer)
        "seats on flight" (:id updated-flight) "at $" price " (< budget of $"
        (:budget customer) ").")
      updated-flight)
    (do
      (log "Customer" (:id customer) "did not find a flight.")
      nil)))

(def finished-processing?
  "Set to true once all customers have been processed, so that sales process
  can end."
  (atom false))

(defn process-customers [customers]
  "Process `customers` one by one."
  (doseq [customer customers]
    (swap! flights
      (fn [flights]
        (if-let [updated-flight (process-customer flights customer)]
          (assoc flights (:id updated-flight) updated-flight)
          flights))))
  (reset! finished-processing? true))

(defn sales-process []
  "The sales process starts and ends sales periods, until `finished-processing?`
  is true."
  (Thread/sleep input/TIME_BETWEEN_SALES)
  (loop []
    (let [discounted-carrier (rand-nth input/carriers)]
      (start-sale discounted-carrier)
      (Thread/sleep input/TIME_OF_SALES)
      (end-sale discounted-carrier))
      (Thread/sleep input/TIME_BETWEEN_SALES)
    (if (not @finished-processing?)
      (recur))))

(defn main []
  (initialize-flights input/flights)
  (let [f1 (future (time (process-customers input/customers)))
        f2 (future (sales-process))]
    @f1
    @f2)
  (println "Flights:")
  (print-flights @flights))

(main)
(shutdown-agents)


src/flight_simulator/input_random.clj
(ns flight-simulator.input-random
  (:gen-class))

(def flights
  [{:id 0
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100 150 0] ; price; # seats available at that price; # seats taken at that price
              [150  50 0]
              [200  50 0]
              [300  50 0]]}
   {:id 1
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100  50 0]
              [200 150 0]
              [320  20 0]
              [240  30 0]]}
   {:id 2
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 3
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 4
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [350  50 0]]}
   {:id 5
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[150 100 0]
              [300 100 0]]}
   {:id 6
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 7
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 8
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 9
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 10
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 11
    :from "BRU" :to "MAD"
    :carrier "Ryanair"
    :pricing [[150 150 0]
              [300  50 0]]}])

(def customers
  (for [id (range 1000)
        :let [{from :from to :to} (rand-nth flights)]]
    {:id     id
     :from   from
     :to     to
     :seats  (+ (rand-int 4) 1)        ; 1-4
     :budget (+ (rand-int 600) 200)})) ; 200-799

(def carriers (distinct (map :carrier flights)))

(def TIME_BETWEEN_SALES 50) ; milliseconds
(def TIME_OF_SALES 10)


src/flight_simulator/input_simple.clj
(ns flight-simulator.input-simple)

(def flights
  [{:id 0
    :from "BRU" :to "ATL"
    :carrier "Delta"
    :pricing [[600 150 0] ; price; # seats available at that price; # seats taken at that price
              [650  50 0]
              [700  50 0]
              [800  50 0]]}
   {:id 1
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[300 150 0]
              [350  50 0]
              [370  20 0]
              [380  30 0]]}
   {:id 2
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 3
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 4
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}])

(def customers
  [{:id  0 :from "BRU" :to "ATL" :seats 5 :budget 700}
   {:id  1 :from "BRU" :to "ATL" :seats 5 :budget 550}
   {:id  2 :from "BRU" :to "LON" :seats 6 :budget 270}
   {:id  3 :from "BRU" :to "ATL" :seats 4 :budget 600}
   {:id  4 :from "BRU" :to "LON" :seats 3 :budget 270}
   {:id  5 :from "BRU" :to "LON" :seats 9 :budget 250}
   {:id  6 :from "BRU" :to "MAD" :seats 5 :budget 200}
   {:id  7 :from "BRU" :to "MAD" :seats 9 :budget 150}
   {:id  8 :from "BRU" :to "LON" :seats 5 :budget 250}
   {:id  9 :from "BRU" :to "ATL" :seats 4 :budget 500}
   {:id 10 :from "BRU" :to "MAD" :seats 1 :budget 180}
   {:id 11 :from "BRU" :to "LON" :seats 2 :budget 320}
   {:id 12 :from "BRU" :to "ATL" :seats 3 :budget 850}
   {:id 13 :from "BRU" :to "ATL" :seats 4 :budget 200}])

(def carriers (distinct (map :carrier flights)))

(def TIME_BETWEEN_SALES 50) ; milliseconds
(def TIME_OF_SALES 10)


src/flight_simulator/input_random_double_flights.clj
(ns flight-simulator.input-random-double-flights
  (:gen-class))

(def flights
  [{:id 0
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100 150 0] ; price; # seats available at that price; # seats taken at that price
              [150  50 0]
              [200  50 0]
              [300  50 0]]}
   {:id 1
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100  50 0]
              [200 150 0]
              [320  20 0]
              [240  30 0]]}
   {:id 2
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 3
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 4
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [350  50 0]]}
   {:id 5
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[150 100 0]
              [300 100 0]]}
   {:id 6
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 7
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 8
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 9
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 10
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 11
    :from "BRU" :to "MAD"
    :carrier "Ryanair"
    :pricing [[150 150 0]
              [300  50 0]]}
   {:id 12
    :from "MAD" :to "LON"
    :carrier "Ryanair"
    :pricing [[100 150 0] ; price; # seats available at that price; # seats taken at that price
              [150  50 0]
              [200  50 0]
              [300  50 0]]}
   {:id 13
    :from "MAD" :to "LON"
    :carrier "Ryanair"
    :pricing [[100  50 0]
              [200 150 0]
              [320  20 0]
              [240  30 0]]}
   {:id 14
    :from "MAD" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 15
    :from "MAD" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 16
    :from "MAD" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [350  50 0]]}
   {:id 17
    :from "MAD" :to "LON"
    :carrier "Ryanair"
    :pricing [[150 100 0]
              [300 100 0]]}
   {:id 18
    :from "LON" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 19
    :from "LON" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 20
    :from "LON" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 21
    :from "LON" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 22
    :from "LON" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 23
    :from "LON" :to "MAD"
    :carrier "Ryanair"
    :pricing [[150 150 0]
              [300  50 0]]}
              ])

(def customers
  (for [id (range 1000)
        :let [{from :from to :to} (rand-nth flights)]]
    {:id     id
     :from   from
     :to     to
     :seats  (+ (rand-int 4) 1)        ; 1-4
     :budget (+ (rand-int 600) 200)})) ; 200-799

(def carriers (distinct (map :carrier flights)))

(def TIME_BETWEEN_SALES 50) ; milliseconds
(def TIME_OF_SALES 10)


src/flight_simulator/input_random_10000_customers.clj
(ns flight-simulator.input-random-10000-customers
  (:gen-class))

(def flights
  [{:id 0
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100 150 0] ; price; # seats available at that price; # seats taken at that price
              [150  50 0]
              [200  50 0]
              [300  50 0]]}
   {:id 1
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100  50 0]
              [200 150 0]
              [320  20 0]
              [240  30 0]]}
   {:id 2
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 3
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 4
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [350  50 0]]}
   {:id 5
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[150 100 0]
              [300 100 0]]}
   {:id 6
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 7
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 8
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 9
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 10
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 11
    :from "BRU" :to "MAD"
    :carrier "Ryanair"
    :pricing [[150 150 0]
              [300  50 0]]}])

(def customers
  (for [id (range 10000)
        :let [{from :from to :to} (rand-nth flights)]]
    {:id     id
     :from   from
     :to     to
     :seats  (+ (rand-int 4) 1)        ; 1-4
     :budget (+ (rand-int 600) 200)})) ; 200-799

(def carriers (distinct (map :carrier flights)))

(def TIME_BETWEEN_SALES 50) ; milliseconds
(def TIME_OF_SALES 10)


src/flight_simulator/input_random_double_flights_10000_customers.clj
(ns flight-simulator.input-random-double-flights-10000-customers
  (:gen-class))

(def flights
  [{:id 0
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100 150 0] ; price; # seats available at that price; # seats taken at that price
              [150  50 0]
              [200  50 0]
              [300  50 0]]}
   {:id 1
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[100  50 0]
              [200 150 0]
              [320  20 0]
              [240  30 0]]}
   {:id 2
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 3
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 4
    :from "BRU" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [350  50 0]]}
   {:id 5
    :from "BRU" :to "LON"
    :carrier "Ryanair"
    :pricing [[150 100 0]
              [300 100 0]]}
   {:id 6
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 7
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 8
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 9
    :from "BRU" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 10
    :from "BRU" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 11
    :from "BRU" :to "MAD"
    :carrier "Ryanair"
    :pricing [[150 150 0]
              [300  50 0]]}
   {:id 12
    :from "MAD" :to "LON"
    :carrier "Ryanair"
    :pricing [[100 150 0] ; price; # seats available at that price; # seats taken at that price
              [150  50 0]
              [200  50 0]
              [300  50 0]]}
   {:id 13
    :from "MAD" :to "LON"
    :carrier "Ryanair"
    :pricing [[100  50 0]
              [200 150 0]
              [320  20 0]
              [240  30 0]]}
   {:id 14
    :from "MAD" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 15
    :from "MAD" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [300  50 0]]}
   {:id 16
    :from "MAD" :to "LON"
    :carrier "Brussels Airlines"
    :pricing [[250 100 0]
              [350  50 0]]}
   {:id 17
    :from "MAD" :to "LON"
    :carrier "Ryanair"
    :pricing [[150 100 0]
              [300 100 0]]}
   {:id 18
    :from "LON" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300 100 0]]}
   {:id 19
    :from "LON" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 20
    :from "LON" :to "MAD"
    :carrier "Iberia"
    :pricing [[200 150 0]
              [250  50 0]
              [300  80 0]
              [350  20 0]]}
   {:id 21
    :from "LON" :to "MAD"
    :carrier "Brussels Airlines"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 22
    :from "LON" :to "MAD"
    :carrier "Iberia"
    :pricing [[250 150 0]
              [300  50 0]]}
   {:id 23
    :from "LON" :to "MAD"
    :carrier "Ryanair"
    :pricing [[150 150 0]
              [300  50 0]]}
              ])

(def customers
  (for [id (range 10000)
        :let [{from :from to :to} (rand-nth flights)]]
    {:id     id
     :from   from
     :to     to
     :seats  (+ (rand-int 4) 1)        ; 1-4
     :budget (+ (rand-int 600) 200)})) ; 200-799

(def carriers (distinct (map :carrier flights)))

(def TIME_BETWEEN_SALES 50) ; milliseconds
(def TIME_OF_SALES 10)


src/flight_simulator/parallel.clj
(ns flight-simulator.parallel
	(:require [clojure.string]
		[clojure.pprint]
		[flight-simulator.input-random :as input]
		;[flight-simulator.input-simple :as input]
		)
	)

(def logger (agent nil))
;(defn log [& msgs] (send logger (fn [_] (apply println msgs))))
(defn log [& msgs] true)

(def flights
	"the flights are stored in a list"
	(atom []))

(def flights-per-carrier
	(atom []))

(def recurcount
	(atom 0))

(def done? (atom false))

(defn init-flights [init-value]
	(reset! flights (map (fn [flight] (ref flight)) init-value))
	)

(defn init-flights-per-carrier [carriers]
	(reset! flights-per-carrier 
		(for [carrier input/carriers] 
			{ :carrier carrier
				:flights (filter 
					(fn [flight] 
						(= (:carrier @flight) carrier)) 
					@flights)})))

(defn flights-of-carrier [carrier]
	(first (map :flights
		(filter (fn [carrierslist] 
			(= (:carrier carrierslist) carrier))
		@flights-per-carrier))))

(defn update-pricing [flight factor]
	"updates the price of a given flight(ref) since we're updating the ref,
	it must be called within a dosync procedure"
	(ref-set flight 
		(update @flight :pricing 
			#(map (fn [[p a t]] [(* p factor) a t]) %)))
	)

(defn update-carrier-price [carrier, factor]
	(let [carrierflights (flights-of-carrier carrier)]
		(dosync
			(doseq [flight carrierflights]
				(update-pricing flight factor))
			))
	)

(defn start-sale [carrier]
	(log carrier "sale started")
	(update-carrier-price carrier 0.80))

(defn end-sale [carrier]
	(log carrier "sale ended")
	(update-carrier-price carrier 1.25))

(defn print-customer [{:keys [id from to seats budget]}]
	"Prints a single customer"
	(println (clojure.pprint/cl-format nil "Customer ~3d From ~a To ~a" id from to)))

(defn print-customers [customers]
	"Prints the customers (to test if the set was successfull)."
	(doseq [customer customers]
		(print-customer customer))
	)


(defn print-flights [flights]
  "Print `flights`."
  (letfn [(pricing->str [pricing]
            (->> pricing
              (map (fn [[p a t]] (clojure.pprint/cl-format nil "$~3d: ~3d ~3d" p a t)))
              (clojure.string/join ", ")))]
    (doseq [flight flights]
    	(let [{:keys [id from to pricing]} @flight]
    	      (println (clojure.pprint/cl-format nil "Flight ~3d from ~a to ~a: ~a"
    	        id from to (pricing->str pricing)))))))


(defn buy-seats [flight, seats, price]
	(dosync
		(let [result (filter (fn [[p a t]] (and (= p price) (>= a seats))) (:pricing @flight))]
		(if (not (empty? result)) 
			(do
				(ref-set flight
					(update @flight :pricing
						#(map (fn [[p a t]] 
							(if (= p price) 
								[p (- a seats) (+ t seats)]
								[p a t])) %)))
				true)
			false))
		)
	)

(defn find-flights-with-n-seats-and-within-budget [from to seats budget] 
	(filter (fn [flight-ref]
		(and
			(= (:from @flight-ref) from)
			(= (:to @flight-ref) to)
			(not (nil? 
				(some (fn [[p a t]]
					(and (< p budget)
						(>= a seats))) (:pricing @flight-ref))
				)
			))) @flights)
	)

(defn cheapest-available-price [flight budget seats]
	(loop [prices (sort-by first (:pricing @flight))]
		(if (empty? prices) nil
		(let [current (first prices)
			  currentprice (first current)
			  currentavailable (nth current 1)]
			  (if (and (< currentprice budget)
			  		   (>= currentavailable seats))
			  currentprice
			  (recur (rest prices))))
		)))

(defn find-cheapest-flight-with-n-seats-and-within-budget [from to seats budget]
	(let [flights (sort-by (fn [flight] (cheapest-available-price flight budget seats)) (find-flights-with-n-seats-and-within-budget from to seats budget))]
		(if (empty? flights)
			nil
			(first flights))
		)
	)

(defn process-customer [{:keys [id from to seats budget]}]
	(let [flight (find-cheapest-flight-with-n-seats-and-within-budget from to seats budget)]

		  (if (nil? flight)
		  	(log "Customer" id "did not find a flight")
		  	(let [price (cheapest-available-price flight budget seats)]
		  		(if (buy-seats flight seats price)
		  			(log "Customer" id "booked" seats "seat(s) for $" price "with budget $" budget "on flight" (:id @flight))
		  			(do (reset! recurcount (+ @recurcount 1)) (recur {:id id :from from :to to :seats seats :budget budget})))
		  		)
		  	)
		))

(defn sales-process []
	"The sales process starts and ends sales periods, until `done?`
	is true."
	(Thread/sleep input/TIME_BETWEEN_SALES)
	(loop []
		(let [discounted-carrier (rand-nth input/carriers)]
			(start-sale discounted-carrier)
			(Thread/sleep input/TIME_OF_SALES)
			(end-sale discounted-carrier))
		(Thread/sleep input/TIME_BETWEEN_SALES)
		(if (not @done?)
			(recur))))

(defn process-customers [customers]
	(doall (pmap process-customer customers))
	(reset! done? true))

(defn -main []
	(init-flights input/flights)
	(init-flights-per-carrier input/carriers)
	(let [sales-loop (future (sales-process))
		  customers-loop (future (time (process-customers input/customers)))]

	; (println (buy-seats (first @flights) 4 (first (first (:pricing @(first @flights))))))
	; (println (first (find-flights-with-n-seats-and-within-budget "BRU" "MAD" 5 1000)))
	; (println (cheapest-available-price 
	; 	(first (find-flights-with-n-seats-and-within-budget "BRU" "MAD" 5 1000))
	; 	5 1000))
	; (println (find-cheapest-flight-with-n-seats-and-within-budget "BRU" "MAD" 5 1000))

	;(process-customer (first input/customers))
	@customers-loop
	@sales-loop
	;(println "----")
	; (print-flights @flights)
	;(println "----")
	(println @recurcount)
	(shutdown-agents)
	)
)
